src/OpenTelemetry/Batch.cs

```diff
@@ -109,6 +140,24 @@ public void Dispose()
             /// <inheritdoc/>
             public bool MoveNext()
             {
+                if (typeof(T) == typeof(Metric))
```

Comment:
@cijothomas's thought was that this would get optimized out when `T != Metric`

Reply:
alternate option to try.
Similar to this:
https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Trace/TracerProviderSdk.cs#L145

Reply:
I haven't yet had the chance to try the alternate approach you've suggested, but before I make those changes here's some perf numbers with and without this `if (typeof(T) == typeof(Metric))` block.

The benchmark simulates what the `SimpleActivityExportProcessor` does in order to assess the impact of the `if (typeof(T) == typeof(Metric))` check.

<details>
  <summary>Benchmark</summary>

```charp
    [MemoryDiagnoser]
    public class BatchBenchmarks
    {
        [Benchmark]
        public void ActivityBatch()
        {
            var activity = new Activity("activity");
            var batch = new Batch<Activity>(activity);

            foreach (var item in batch)
            {
                item.Start();
                item.Stop();
            }
        }
    }
```

</details>

**With `if (typeof(T) == typeof(Metric))`**

|        Method |     Mean |   Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------- |---------:|--------:|--------:|-------:|------:|------:|----------:|
| ActivityBatch | 493.1 ns | 5.68 ns | 5.04 ns | 0.0486 |     - |     - |     408 B |

**Without `if (typeof(T) == typeof(Metric))`**

|        Method |     Mean |   Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------- |---------:|--------:|--------:|-------:|------:|------:|----------:|
| ActivityBatch | 490.6 ns | 6.43 ns | 5.70 ns | 0.0486 |     - |     - |     408 B |


I've gotten some pretty varied results. Sometimes the `Without` run is slower than the `With` run.

Reply:
I tried similar benchmarks and results are not consistent. Sometimes the new Batch is fast! 
I think the ActivityCreation itself is ~450-500. And the additional overhead we introduced is within margin of error.

```
namespace Benchmarks.Metrics
{
    [MemoryDiagnoser]
    public class MetricCollectBenchmarks
    {
        private double totalMs;

        [GlobalSetup]
        public void Setup()
        {
        }

        [GlobalCleanup]
        public void Cleanup()
        {
        }

        [Benchmark]
        public void BatchOriginal()
        {
            var activity = new Activity("activity");
            activity.Start();
            activity.Stop();
            var batch = new BatchOriginal<Activity>(activity);
            foreach (var act in batch)
            {
                var ms = act.Duration.TotalMilliseconds;
                this.totalMs += ms;
            }
        }

        [Benchmark]
        public void BatchNew()
        {
            var activity = new Activity("activity");
            activity.Start();
            activity.Stop();
            var batch = new Batch<Activity>(activity);
            foreach (var act in batch)
            {
                var ms = act.Duration.TotalMilliseconds;
                this.totalMs += ms;
            }
        }
    }
}
```

Run1:

|        Method |     Mean |    Error |   StdDev |   Median |  Gen 0 | Allocated |
|-------------- |---------:|---------:|---------:|---------:|-------:|----------:|
| BatchOriginal | 463.9 ns | 15.78 ns | 44.50 ns | 461.6 ns | 0.0916 |     384 B |
|      BatchNew | 442.1 ns | 18.27 ns | 52.12 ns | 423.8 ns | 0.0916 |     384 B |

Run2:
|        Method |     Mean |    Error |   StdDev |  Gen 0 | Allocated |
|-------------- |---------:|---------:|---------:|-------:|----------:|
| BatchOriginal | 476.8 ns | 12.50 ns | 35.47 ns | 0.0973 |     408 B |
|      BatchNew | 469.8 ns |  9.40 ns | 22.34 ns | 0.0973 |     408 B |

Reply:
@reyang The benchmarks doesn't show any consistently measurable regression (sometimes new one is faster).

Given that adding metrics to Batch is not affecting other signals in measurable way, we should be okay to add metrics support to Batch and achieve cleaner exporter code. 

Please let us know your comments.

Reply:
+1.